Yank/restraints.py

```diff
@@ -551,20 +698,295 @@ def _determineBondParameters(self):
+        self._determine_restraint_parameters(state, system, positions)
+
+    def _is_collinear(self, positions, atoms, THRESHOLD=0.9):
+        """Report whether any sequential vectors in a sequence of atoms are collinear to within a given dot product threshold.
+
+        Parameters
+        ----------
+        positions : simtk.unit.Quantity of natoms x 3 with units compatible with nanometers
+            Reference positions to use for imposing restraints
+        atoms : array-like, dtype:int, length:natoms
+            A seris of indices of atoms
+        THRESHOLD : float, optional, default=0.9
+            Atoms are not colinear if their sequential vector separation dot products are less than THRESHOLD
+
+        Returns
+        -------
+        result : bool
+            Returns True if any sequential pair of vectors is collinear; False otherwise.
+
+        """
+        result = False
+        for i in range(len(atoms)-2):
+            v1 = positions[atoms[i+1],:] - positions[atoms[i],:]
+            v2 = positions[atoms[i+2],:] - positions[atoms[i+1],:]
+            innerprod = np.dot(v1, v2)
+            result = result or (innerprod > THRESHOLD)
+
+        return result
+
+    def _select_restraint_atoms(self, positions, receptor_atoms, ligand_atoms):
+        """
+        Select atoms to be used in restraint.
+
+        Parameters
+        ----------
+        positions : simtk.unit.Quantity of natoms x 3 with units compatible with nanometers
+           Reference positions to use for imposing restraints
+        receptor_atoms : list of int
+            List of atom indices belonging to the receptor
+        ligand_atoms : list of int
+            List of atom indices belonging to the ligand
+
+        Returns
+        -------
+        restraint_atoms : list of int
+            List of six atom indices used in the restraint.
+            restraint_atoms[0:3] belong to the receptor
+            restraint_atoms[4:6] belong to the ligand
+
+        Notes
+        -----
+        The current algorithm simply selects random subsets of receptor and ligand atoms
+        and rejects those that are too close to collinear. Future updates can further refine
+        this algorithm.
+
+        """
+        if (len(receptor_atoms) < 3) or (len(ligand_atoms) < 3):
+            raise Exception('Both receptor_atoms (len %d) and ligand_atoms (len %d) must contain at least three atoms.' % (len(receptor_atoms), len(ligand_atoms)))
+
+        # Iterate until we have found a set of non-collinear atoms.
+        accepted = False
+        while (not accepted):
+            # Select three random atoms from the receptor and ligand
+            restraint_atoms = random.sample(receptor_atoms, 3) + random.sample(ligand_atoms, 3)
+            # Reject collinear sets of atoms.
+            accepted = not self._is_collinear(positions, restraint_atoms)
+
+        return restraint_atoms
+
+    def _determine_restraint_parameters(self, state, system, positions):
```

Comment:
I see, thanks for clarifying. `positions` is not used anyway because you already have `self.positions`.